<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns='http://www.w3.org/1999/xhtml' xml:lang='en' lang='en'>
<head>
<title>poky - Poky Build Tool and Metadata</title>
<meta name='generator' content='cgit v0.9.1'/>
<meta name='robots' content='index, nofollow'/>
<link rel='stylesheet' type='text/css' href='/cgit/cgit.css'/>
<link rel='shortcut icon' href='/favicon.ico'/>
<link rel='alternate' title='Atom feed' href='http://git.yoctoproject.org/cgit/cgit.cgi/poky/atom/meta/recipes-graphics/wayland/weston/weston-launch-shell.patch?h=master' type='application/atom+xml'/>
</head>
<body>
<div id='cgit'><table id='header'>
<tr>
<td class='logo' rowspan='2'><a href='/cgit/cgit.cgi/'><img src='/cgit/yocto-project-transp.png' alt='cgit logo'/></a></td>
<td class='main'><a href='/cgit/cgit.cgi/'>index</a> : <a title='poky' href='/cgit/cgit.cgi/poky/'>poky</a></td><td class='form'><form method='get' action=''>
<select name='h' onchange='this.form.submit();'>
<option value='1.1_M1'>1.1_M1</option>
<option value='1.1_M2'>1.1_M2</option>
<option value='1.1_M3'>1.1_M3</option>
<option value='1.1_M4'>1.1_M4</option>
<option value='1.2_M1'>1.2_M1</option>
<option value='1.2_M2'>1.2_M2</option>
<option value='1.2_M3'>1.2_M3</option>
<option value='1.2_M4'>1.2_M4</option>
<option value='1.3_M1'>1.3_M1</option>
<option value='1.3_M2'>1.3_M2</option>
<option value='1.3_M3'>1.3_M3</option>
<option value='1.3_M4'>1.3_M4</option>
<option value='1.3_M5'>1.3_M5</option>
<option value='1.3_beta'>1.3_beta</option>
<option value='1.4_M1'>1.4_M1</option>
<option value='1.4_M3'>1.4_M3</option>
<option value='1.4_M4'>1.4_M4</option>
<option value='1.4_M5'>1.4_M5</option>
<option value='bernard'>bernard</option>
<option value='blinky'>blinky</option>
<option value='clyde'>clyde</option>
<option value='danny'>danny</option>
<option value='danny-next'>danny-next</option>
<option value='denzil'>denzil</option>
<option value='dylan'>dylan</option>
<option value='edison'>edison</option>
<option value='elroy'>elroy</option>
<option value='green'>green</option>
<option value='laverne'>laverne</option>
<option value='master' selected='selected'>master</option>
<option value='master-next'>master-next</option>
<option value='pinky'>pinky</option>
<option value='purple'>purple</option>
</select> <input type='submit' name='' value='switch'/></form></td></tr>
<tr><td class='sub'>Poky Build Tool and Metadata</td><td class='sub right'>git repository hosting</td></tr></table>
<table class='tabs'><tr><td>
<a href='/cgit/cgit.cgi/poky/'>summary</a><a href='/cgit/cgit.cgi/poky/refs/'>refs</a><a href='/cgit/cgit.cgi/poky/log/meta/recipes-graphics/wayland/weston/weston-launch-shell.patch'>log</a><a class='active' href='/cgit/cgit.cgi/poky/tree/meta/recipes-graphics/wayland/weston/weston-launch-shell.patch'>tree</a><a href='/cgit/cgit.cgi/poky/commit/meta/recipes-graphics/wayland/weston/weston-launch-shell.patch'>commit</a><a href='/cgit/cgit.cgi/poky/diff/meta/recipes-graphics/wayland/weston/weston-launch-shell.patch'>diff</a><a href='/cgit/cgit.cgi/poky/stats/meta/recipes-graphics/wayland/weston/weston-launch-shell.patch'>stats</a></td><td class='form'><form class='right' method='get' action='/cgit/cgit.cgi/poky/log/meta/recipes-graphics/wayland/weston/weston-launch-shell.patch'>
<select name='qt'>
<option value='grep'>log msg</option>
<option value='author'>author</option>
<option value='committer'>committer</option>
<option value='range'>range</option>
</select>
<input class='txt' type='text' size='10' name='q' value=''/>
<input type='submit' value='search'/>
</form>
</td></tr></table>
<div class='path'>path: <a href='/cgit/cgit.cgi/poky/tree/'>root</a>/<a href='/cgit/cgit.cgi/poky/tree/meta'>meta</a>/<a href='/cgit/cgit.cgi/poky/tree/meta/recipes-graphics'>recipes-graphics</a>/<a href='/cgit/cgit.cgi/poky/tree/meta/recipes-graphics/wayland'>wayland</a>/<a href='/cgit/cgit.cgi/poky/tree/meta/recipes-graphics/wayland/weston'>weston</a>/<a href='/cgit/cgit.cgi/poky/tree/meta/recipes-graphics/wayland/weston/weston-launch-shell.patch'>weston-launch-shell.patch</a></div><div class='content'>blob: ffe0608ee27c2c1278ee64a027db65a67d8d3db0 (<a href='/cgit/cgit.cgi/poky/plain/meta/recipes-graphics/wayland/weston/weston-launch-shell.patch'>plain</a>)
<table summary='blob content' class='blob'>
<tr><td class='linenumbers'><pre><a class='no' id='n1' name='n1' href='#n1'>1</a>
<a class='no' id='n2' name='n2' href='#n2'>2</a>
<a class='no' id='n3' name='n3' href='#n3'>3</a>
<a class='no' id='n4' name='n4' href='#n4'>4</a>
<a class='no' id='n5' name='n5' href='#n5'>5</a>
<a class='no' id='n6' name='n6' href='#n6'>6</a>
<a class='no' id='n7' name='n7' href='#n7'>7</a>
<a class='no' id='n8' name='n8' href='#n8'>8</a>
<a class='no' id='n9' name='n9' href='#n9'>9</a>
<a class='no' id='n10' name='n10' href='#n10'>10</a>
<a class='no' id='n11' name='n11' href='#n11'>11</a>
<a class='no' id='n12' name='n12' href='#n12'>12</a>
<a class='no' id='n13' name='n13' href='#n13'>13</a>
<a class='no' id='n14' name='n14' href='#n14'>14</a>
<a class='no' id='n15' name='n15' href='#n15'>15</a>
<a class='no' id='n16' name='n16' href='#n16'>16</a>
<a class='no' id='n17' name='n17' href='#n17'>17</a>
<a class='no' id='n18' name='n18' href='#n18'>18</a>
<a class='no' id='n19' name='n19' href='#n19'>19</a>
<a class='no' id='n20' name='n20' href='#n20'>20</a>
<a class='no' id='n21' name='n21' href='#n21'>21</a>
<a class='no' id='n22' name='n22' href='#n22'>22</a>
<a class='no' id='n23' name='n23' href='#n23'>23</a>
<a class='no' id='n24' name='n24' href='#n24'>24</a>
<a class='no' id='n25' name='n25' href='#n25'>25</a>
<a class='no' id='n26' name='n26' href='#n26'>26</a>
<a class='no' id='n27' name='n27' href='#n27'>27</a>
<a class='no' id='n28' name='n28' href='#n28'>28</a>
<a class='no' id='n29' name='n29' href='#n29'>29</a>
<a class='no' id='n30' name='n30' href='#n30'>30</a>
<a class='no' id='n31' name='n31' href='#n31'>31</a>
<a class='no' id='n32' name='n32' href='#n32'>32</a>
<a class='no' id='n33' name='n33' href='#n33'>33</a>
<a class='no' id='n34' name='n34' href='#n34'>34</a>
<a class='no' id='n35' name='n35' href='#n35'>35</a>
<a class='no' id='n36' name='n36' href='#n36'>36</a>
<a class='no' id='n37' name='n37' href='#n37'>37</a>
<a class='no' id='n38' name='n38' href='#n38'>38</a>
<a class='no' id='n39' name='n39' href='#n39'>39</a>
<a class='no' id='n40' name='n40' href='#n40'>40</a>
<a class='no' id='n41' name='n41' href='#n41'>41</a>
<a class='no' id='n42' name='n42' href='#n42'>42</a>
<a class='no' id='n43' name='n43' href='#n43'>43</a>
<a class='no' id='n44' name='n44' href='#n44'>44</a>
<a class='no' id='n45' name='n45' href='#n45'>45</a>
<a class='no' id='n46' name='n46' href='#n46'>46</a>
<a class='no' id='n47' name='n47' href='#n47'>47</a>
<a class='no' id='n48' name='n48' href='#n48'>48</a>
<a class='no' id='n49' name='n49' href='#n49'>49</a>
<a class='no' id='n50' name='n50' href='#n50'>50</a>
<a class='no' id='n51' name='n51' href='#n51'>51</a>
<a class='no' id='n52' name='n52' href='#n52'>52</a>
<a class='no' id='n53' name='n53' href='#n53'>53</a>
<a class='no' id='n54' name='n54' href='#n54'>54</a>
<a class='no' id='n55' name='n55' href='#n55'>55</a>
<a class='no' id='n56' name='n56' href='#n56'>56</a>
<a class='no' id='n57' name='n57' href='#n57'>57</a>
<a class='no' id='n58' name='n58' href='#n58'>58</a>
<a class='no' id='n59' name='n59' href='#n59'>59</a>
<a class='no' id='n60' name='n60' href='#n60'>60</a>
<a class='no' id='n61' name='n61' href='#n61'>61</a>
<a class='no' id='n62' name='n62' href='#n62'>62</a>
<a class='no' id='n63' name='n63' href='#n63'>63</a>
<a class='no' id='n64' name='n64' href='#n64'>64</a>
<a class='no' id='n65' name='n65' href='#n65'>65</a>
<a class='no' id='n66' name='n66' href='#n66'>66</a>
<a class='no' id='n67' name='n67' href='#n67'>67</a>
<a class='no' id='n68' name='n68' href='#n68'>68</a>
<a class='no' id='n69' name='n69' href='#n69'>69</a>
<a class='no' id='n70' name='n70' href='#n70'>70</a>
<a class='no' id='n71' name='n71' href='#n71'>71</a>
<a class='no' id='n72' name='n72' href='#n72'>72</a>
<a class='no' id='n73' name='n73' href='#n73'>73</a>
<a class='no' id='n74' name='n74' href='#n74'>74</a>
</pre></td>
<td class='lines'><pre><code>Upstream-Status: Backport
Signed-off-by: Ross Burton &lt;ross.burton@intel.com&gt;

From ff3230952a68077669e0ea0ac3ceb234273556fc Mon Sep 17 00:00:00 2001
From: Quentin Glidic &lt;sardemff7+git@sardemff7.net&gt;
Date: Fri, 17 May 2013 16:20:37 +0200
Subject: [PATCH] weston-launch: Run weston in the user login shell

This patch brings back the user environment from the shell.
In the future, weston-launch could create the Wayland socket earlier, in
which case the user's shell could be used to run Wayland-specific tools
in the new Weston session.

Signed-off-by: Quentin Glidic &lt;sardemff7+git@sardemff7.net&gt;
---
 src/weston-launch.c |   19 ++++++++++++++++---
 1 file changed, 16 insertions(+), 3 deletions(-)

diff --git a/src/weston-launch.c b/src/weston-launch.c
index 64d4a8a..89c3c5a 100644
--- a/src/weston-launch.c
+++ b/src/weston-launch.c
@@ -60,6 +60,8 @@
 
 #include "weston-launch.h"
 
+#define MAX_ARGV_SIZE 256
+
 struct weston_launch {
 	struct pam_conv pc;
 	pam_handle_t *ph;
@@ -523,8 +525,9 @@ main(int argc, char *argv[])
 	struct weston_launch wl;
 	char **env;
 	int i, c;
-	char **child_argv;
+	char *child_argv[MAX_ARGV_SIZE];
 	char *tty = NULL, *new_user = NULL;
+	char *term;
 	int sleep_fork = 0;
 	struct option opts[] = {
 		{ "user",    required_argument, NULL, 'u' },
@@ -562,8 +565,8 @@ main(int argc, char *argv[])
 		}
 	}
 
-	child_argv = &amp;argv[optind-1];
-	child_argv[0] = BINDIR "/weston";
+	if ((argc - optind) &gt; (MAX_ARGV_SIZE - 5))
+		error(1, E2BIG, "Too many arguments to pass to weston");
 
 	if (new_user)
 		wl.pw = getpwnam(new_user);
@@ -572,7 +575,17 @@ main(int argc, char *argv[])
 	if (wl.pw == NULL)
 		error(1, errno, "failed to get username");
 
+	child_argv[0] = wl.pw-&gt;pw_shell;
+	child_argv[1] = "-l";
+	child_argv[2] = "-c";
+	child_argv[3] = BINDIR "/weston \"$@\"";
+	child_argv[4] = "weston";
+	for (i = 0; i &lt; (argc - optind); ++i)
+		child_argv[5+i] = argv[optind+i];
+
+	term = getenv("TERM");
 	clearenv();
+	setenv("TERM", term, 1);
 	setenv("USER", wl.pw-&gt;pw_name, 1);
 	setenv("LOGNAME", wl.pw-&gt;pw_name, 1);
 	setenv("HOME", wl.pw-&gt;pw_dir, 1);
-- 
1.7.10.4

</code></pre></td></tr></table>
</div> <!-- class=content -->
<div class='footer'>generated  by cgit v0.9.1 at 2013-09-17 12:41:36 (GMT)</div>
</div> <!-- id=cgit -->
</body>
</html>
